### PR TITLE
Add pricing for BRL swaps

### PR DESCRIPTION
**Pull request by r-weeks**

### CODE REVIEW

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/DispatchingRateComputationFn.java [hunk 1]
```diff
@@ -246,6 +269,12 @@ public double explainRate(
       builder.put(ExplainKey.FIXED_RATE, rate);
       builder.put(ExplainKey.COMBINED_RATE, rate);
       return rate;
+    } else if (computation instanceof FixedOvernightCompoundedAnnualRateComputation) {
+      // inline code (performance) avoiding need for FixedRateComputationFn implementation
+      double rate = ((FixedOvernightCompoundedAnnualRateComputation) computation).getRate();
+      builder.put(ExplainKey.FIXED_RATE, rate);
+      builder.put(ExplainKey.COMBINED_RATE, rate);
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/DispatchingRateComputationFn.java]: Should the COMBINED_RATE be the result of `getSimpleRate`?

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFn.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+import com.opengamma.strata.pricer.rate.RatesProvider;
+import com.opengamma.strata.product.rate.OvernightCompoundedAnnualRateComputation;
+
+/**
+ * Rate computation implementation for a rate based on a single overnight index that is compounded using an annual rate.
+ * <p>
+ * Rates that are already fixed are retrieved from the time series of the {@link RatesProvider}.
+ * Rates that are in the future and not in the cut-off period are computed as unique forward rate
+ * in the full future period.
+ * Rates that are in the cut-off period (already fixed or forward) are compounded.
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFn.java]: There is no "cut-off period", right?

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/swap/DiscountingSwapProductPricer.java [hunk 3]
```diff
@@ -371,6 +418,18 @@ public PointSensitivityBuilder parRateSensitivity(ResolvedSwap swap, RatesProvid
...
    */
   public PointSensitivityBuilder parSpreadSensitivity(ResolvedSwap swap, RatesProvider provider) {
+    // does the fixed leg of the swap, if it exists, have a future value notional
+    if (!swap.getLegs(SwapLegType.FIXED).isEmpty()) {
+      ResolvedSwapLeg fixedLeg = fixedLeg(swap);
+      SwapPaymentPeriod firstPeriod = fixedLeg.getPaymentPeriods().get(0);
+      if (firstPeriod instanceof RatePaymentPeriod) {
+        RatePaymentPeriod payment = (RatePaymentPeriod) firstPeriod;
+        RateAccrualPeriod firstAccrualPeriod = payment.getAccrualPeriods().get(0);
+        if (firstAccrualPeriod.getRateComputation() instanceof FixedOvernightCompoundedAnnualRateComputation) {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/swap/DiscountingSwapProductPricer.java]: Extract the logic for getting `FixedOvernightCompoundedAnnualRateComputation` to a method as it exists in multiple places in this file:

`public Optonal<FixedOvernightCompoundedAnnualRateComputation> findFixedOvernightCompoundedAnnualRateComputation();`

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+import com.opengamma.strata.product.swap.ResolvedSwap;
+import com.opengamma.strata.product.swap.ResolvedSwapTrade;
+import com.opengamma.strata.product.swap.SwapTrade;
+import com.opengamma.strata.product.swap.type.FixedOvernightSwapConvention;
+import com.opengamma.strata.product.swap.type.FixedOvernightSwapTemplate;
+import com.opengamma.strata.product.swap.type.FixedRateSwapLegConvention;
+import com.opengamma.strata.product.swap.type.ImmutableFixedOvernightSwapConvention;
+import com.opengamma.strata.product.swap.type.OvernightRateSwapLegConvention;
+
+
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java]: Only one blank line

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+  
+  public static final FixedOvernightSwapConvention BRL_OIS_CONVENTION = ImmutableFixedOvernightSwapConvention.of(
+      "BRL-FIXED-TERM-CDI-OIS",
+      BRL_FIXED_CONV,
+      BRL_FLOATING_CONVENTION,
+      DaysAdjustment.ofBusinessDays(0, BRL_CDI.getFixingCalendar()));
+  
+  /** Market values for the BRL curve */
+  private static final double[] OIS_MARKET_QUOTES = new double[] {
+      0.1380, 0.1340, 0.1250, 0.1175, 0.1150,
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java]: These should be on one line

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+      0.1150, 0.1160};
+  private static final int OIS_NB_NODES = OIS_MARKET_QUOTES.length;
+  private static final String[] OIS_ID_VALUE = new String[] {
+      "OIS-3M", "OIS-6M", "OIS-1Y", "OIS-2Y", "OIS-3Y",
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java]: These should be on one line

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    groupNodes.add(ALL_NODES);
+    CURVES_NODES.add(groupNodes);
+  }
+  
+  /** All metadata by groups */
+  private static final List<List<CurveMetadata>> CURVES_METADATA = new ArrayList<>();
+  static {
+    List<CurveMetadata> groupMetadata = new ArrayList<>();
+    groupMetadata.add(DefaultCurveMetadata.builder().curveName(ALL_CURVE_NAME).xValueType(ValueType.YEAR_FRACTION)
+        .yValueType(ValueType.ZERO_RATE).dayCount(CURVE_DC).build());
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java]: When formatting a long builder, put every method call after `builder()` on a new line

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    CURVES_METADATA.add(groupMetadata);
+  }
+  
+  private static final DiscountingSwapProductPricer SWAP_PRICER =
+      DiscountingSwapProductPricer.DEFAULT;
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java]: Avoid breaking a line in two where not necessary

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    double shift = 1.0E-6;
+    Function<MarketData, RatesProvider> f =
+        marketData -> CALIBRATOR.calibrate(CURVE_GROUP_DEFN, marketData, REF_DATA);
+    calibration_market_quote_sensitivity_check(f, shift);
+  }
+  
+  //-------------------------------------------------------------------------
+  private void calibration_market_quote_sensitivity_check(
+      Function<MarketData, RatesProvider> calibrator,
+      double shift) {
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/curve/CalibrationDiscountingBrlTest.java]: When method parameters are split onto multiple lines, there should be a blank line after the `{`

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFnTest.java [hunk 10]
```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (C) 2018 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+/*
+ * Copyright (C) 2018 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFnTest.java]: Two licenses :-)

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFnTest.java [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+      OvernightIndexObservation.of(BRL_CDI, date(2015, 1, 8), REF_DATA),
+      OvernightIndexObservation.of(BRL_CDI, date(2015, 1, 9), REF_DATA),
+      OvernightIndexObservation.of(BRL_CDI, date(2015, 1, 12), REF_DATA),
+      OvernightIndexObservation.of(BRL_CDI, date(2015, 1, 13), REF_DATA),
+      OvernightIndexObservation.of(BRL_CDI, date(2015, 1, 14), REF_DATA),
+      OvernightIndexObservation.of(BRL_CDI, date(2015, 1, 15), REF_DATA)};
+
+  private static final double[] FIXING_RATES = {
+      0.0012, 0.0023, 0.0034,
+      0.0045, 0.0056, 0.0067, 0.0078};
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFnTest.java]: No need to split these lines

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFnTest.java [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+  private static final double[] FORWARD_RATES = {
+      0.0112, 0.0123, 0.0134,
```
Comment (jodastephen) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFnTest.java]: No need to split these lines

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFn.java (hunk 2 above)
Comment (r-weeks) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/ForwardOvernightCompoundedAnnualRateComputationFn.java]: No. will remove any reference.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/DispatchingRateComputationFn.java (hunk 1 above)
Comment (r-weeks) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/DispatchingRateComputationFn.java]: Yes I was wondering about this. The simple rate or the accrued rate?

Comment (jodastephen) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/rate/DispatchingRateComputationFn.java]: FIXED_RATE is the "normal" rate in the contract, COMBINED_RATE is the rate after processing (gearing, spread etc.)

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/swap/DiscountingSwapProductPricer.java [hunk 13]
```diff
@@ -360,28 +355,25 @@ public PointSensitivityBuilder parRateSensitivity(ResolvedSwap swap, RatesProvid
...
-        double otherLegsConvertedPvBar = -Math.pow(-otherLegsConvertedPv  / (notional * df) + 1.0d, 1.0 / af - 1.0d) / (af * notional * df);
-        double dfBar = Math.pow(-otherLegsConvertedPv  / (notional * df) + 1.0d, 1.0 / af - 1.0d) * otherLegsConvertedPv / (af * notional * df * df);
-        PointSensitivityBuilder otherLegsConvertedPvDr = PointSensitivityBuilder.none();
-        for (ResolvedSwapLeg leg : swap.getLegs()) {
-          if (leg != fixedLeg) {
-            PointSensitivityBuilder pvLegDr = getLegPricer().presentValueSensitivity(leg, provider)
-                .multipliedBy(provider.fxRate(leg.getCurrency(), ccyFixedLeg));
-            otherLegsConvertedPvDr = otherLegsConvertedPvDr.combinedWith(pvLegDr);
-          }
+    Optional<FixedOvernightCompoundedAnnualRateComputation> annualRateCompIfPresent = findFixedOvernightCompoundedAnnualRateComputation(fixedLeg);
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/swap/DiscountingSwapProductPricer.java]: I normally call variables of type `Optional` something like `xxxOpt`.

**Pull request merged**